onnxruntime/core/providers/webnn/builders/impl/concat_op_builder.cc

```diff
@@ -52,6 +56,38 @@ Status ConcatOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
   return Status::OK();
 }
 
+bool ConcatOpBuilder::HasSupportedInputsImpl(const Node& node, const emscripten::val& wnn_limits,
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& op_type = node.OpType();
+  int32_t input0_type;
+
+  if (!GetType(*input_defs[0], input0_type, logger))
+    return false;
+
+  if (!IsSupportedDataType(input0_type, wnn_limits["concat"]["inputs"]["dataTypes"])) {
+    LOGS(logger, VERBOSE) << "[" << op_type
+                          << "] Input type: [" << input0_type
+                          << "] is not supported for now";
+    return false;
+  }
+
+  for (size_t i = 1; i < input_defs.size(); i++) {
+    int32_t input_type;
+    if (!GetType(*input_defs[i], input_type, logger)) {
+      return false;
+    }
+
+    if (input0_type != input_type) {
```

Comment:
I think this could simply be an error rather than returning false, as concat requires all inputs to be the same data type. Even if we fall back to the CPU EP, the CPU EP can't do anything about it either - it will also fail. Though I suppose it's the existing pattern for this method, to return false rather than throw an error. 🤷‍♂️ (resolve me)

Reply:
I didn't find ONNX's spec restrict the inputs of `Concat` to be same type. https://onnx.ai/onnx/operators/onnx__Concat.html